src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/MSBuildRestoreUtility.cs

```diff
@@ -951,5 +940,237 @@ public static Task ReplayWarningsAndErrorsAsync(IEnumerable<IAssetsLogMessage> m
 
             return logger.LogMessagesAsync(logMessages);
         }
+
+        /// <summary>
+        /// Merge the information from the PackageVersion items to the PackageReferences items.
+        /// </summary>
+        /// <param name="packageReferences"></param>
+        /// <param name="centralPackageVersions"></param>
+        /// <returns></returns>
+        static IEnumerable<IMSBuildItem> MergeCentralPackageVersionsInPackageReferences(IList<IMSBuildItem> packageReferences,
+            IList<IMSBuildItem> centralPackageVersions)
+        {
+            if (packageReferences == null)
+            {
+                throw new ArgumentNullException(nameof(packageReferences));
+            }
+
+            if (centralPackageVersions == null)
+            {
+                throw new ArgumentNullException(nameof(centralPackageVersions));
+            }
+
+            // Do not touch the IsImplicitlyDefined package references. 
+            var explicitlyReferencedPackages = packageReferences
+                   .Where(item => !IsPropertyTrue(item, "IsImplicitlyDefined"))
+                   .ToList();
+
+            var implicitlyReferencedPackages = packageReferences
+                   .Where(item => IsPropertyTrue(item, "IsImplicitlyDefined"))
+                   .ToList();
```

Comment:
Consider moving this to just before it's used. Also, it's only enumerated once, so you don't need the `ToList`, that just adds extra allocations and CPU work.

Reply:
Removed.